README.md
# vscode-jscharting

This extension provides [JSCharting](https://jscharting.com/) chart preview integration for Visual Studio Code.

**JSCharting** is a JavaScript data visualization library offering seamless usage with Vue across all devices and platforms. Every JSCharting license includes a full suite of 150+ chart types including standards such as pie charts, line charts, donut and bar charts. In addition, advanced chart types including Gantt charts, JavaScript Org Charts, interactive charts for stock and finance, seamless grid and calendar charts, JavaScript maps, sparklines, and micro charts all for no additional charge. JSCharting has all the features you need and many you don't yet know you want.


Example Charts: 
[Chart Types](https://jscharting.com/examples/chart-types/)
| [Feature Examples](https://jscharting.com/examples/chart-features/)

[<img src="https://jscharting.com/images/thumbs/Javascript_StackedSalesBars_Chart.png" width="210">](https://jscharting.com/examples/chart-types/bar/ "Bar Charts")
[<img src="https://jscharting.com/images/thumbs/Javascript_UncertaintyFanChart_Chart.png" width="210">](https://jscharting.com/examples/chart-types/area/ "Area Charts")
[<img src="https://jscharting.com/images/thumbs/Javascript_MapWithPies_Chart.png" width="210">](https://jscharting.com/examples/chart-types/geographic-map/ "Mapping")
[<img src="https://jscharting.com/images/thumbs/Javascript_GanttMilestones_Chart.png" width="210">](https://jscharting.com/examples/chart-types/gantt/ "Gantt Charts")

## Features

A preview is available for `*.jsc.json` and `*.jsc.json5` extensions that contain chart configurations. 

![JSON5 Preivew](https://raw.githubusercontent.com/jscharting/vscode-jscharting/master/images/json5preview.png
"Chart Preview")

JSCharting configuration can also be used in md files with `jscharting` codeblock to render charts in preview mode.

# Usage

Run the `JSCharting: Preview Chart` command from `View > Command Palette...` or click the preview icon ![JSON5 Preivew](https://raw.githubusercontent.com/jscharting/vscode-jscharting/master/images/preview_icon.png) in the editor to preview the chart.

## Samples and Source 

Visit the [vscode-jscharting GitHub page](https://github.com/jscharting/vscode-jscharting) for example JSON files and the source code of this extension.  

To preview charts embedded in MD documents, wrap your JSON chart options using the following syntax and preview the MD document.

~~~
```jscharting
{ 
    type:'column',
    series:[{
        points: [ ['A',5], ['B', 6] ]   
    }]
}
```

~~~

Visit the JSCharting [getting started](https://jscharting.com/tutorials/creating-js-charts/) page for more information on chart options.


CHANGELOG.md
# Change Log


## 0.0.3 - 2020.11.13

- Update logo
- Modify package details
- Add license file
- Handle json syntax errors

## 0.0.1 - 2020.11.18

- Initial release

main.js
function main() {
	var jschartingElements = document.querySelectorAll('.jscharting');

	for (const jschartingElement of jschartingElements) {
		try {
			var jscConfig = getConfigFromElement(jschartingElement);
			JSC.Chart(jschartingElement, fillData(jscConfig));
		} catch (e) {
			jschartingElement.innerText = e;
		}
	}
}

function getConfigFromElement(element) {
	if (!element) {
		return null;
	}

	const jsonString = element.textContent;
	element.innerHTML = '';

	return JSON5.parse(jsonString);
}

window.addEventListener(
	"load",
	function () { main(); },
	false
);


function fillData(cfg){
	if(!cfg.series || cfg.series.length ===0){
		cfg.series = [
			{
				name:'Temporary Data',
				points:[
					{name:'A',y:20},
					{name:'B',y:50},
					{name:'C',y:150},
					{name:'D',y:75},
				]
			}
		];
	}
	return cfg;
}

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import * as path from 'path';

const json5 = require('json5');

const supportedExtensions = ['.json', '.json5', '.jsc'];

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	let panel: vscode.WebviewPanel | null;
	context.subscriptions.push(
		vscode.commands.registerCommand('jscharting.start', () => {
			const columnToShowIn = vscode.window.activeTextEditor
				? vscode.window.activeTextEditor.viewColumn
				: undefined;

			if (!columnToShowIn) {
				vscode.window.showWarningMessage('There is no configuration to process!');
				return;
			}

			const document = vscode.window.activeTextEditor?.document;
			if (!document) {
				return;
			}

			const uri = document.uri;
			if (uri.scheme !== 'file') {
				vscode.window.showWarningMessage(`Document scheme ${uri.scheme} not supported!`);
				return;
			}

			const fileName = path.basename(uri.path);
			const extName = path.extname(uri.path);
			const support = !!supportedExtensions.find(e => e.toLowerCase() === extName.toLowerCase());
			if (!support) {
				vscode.window.showWarningMessage(`File ${extName} not supported!`);
				return;
			}

			if (panel) {
				panel.reveal(columnToShowIn);
			} else {
				panel = vscode.window.createWebviewPanel(
					'jscharting',
					`${fileName} preview`,
					vscode.ViewColumn.One,
					{
						enableScripts: true
					}
				);
			}

			const json5OnDiskPath = vscode.Uri.file(
				path.join(context.extensionPath, 'node_modules', 'json5', 'dist/index.min.js')
			);
			const jschartingOnDiskPath = vscode.Uri.file(
				path.join(context.extensionPath, 'node_modules', 'jscharting', 'dist/jscharting.js')
			);

			const json5Src = panel.webview.asWebviewUri(json5OnDiskPath);
			const jschartingSrc = panel.webview.asWebviewUri(jschartingOnDiskPath);

			const draw = () => {
				if (!panel) {
					return;
				}

				const codeText = document.getText();

				let jsonVal, skipUpdate;
				try {
					jsonVal = json5.parse(codeText);
					jsonVal = fillData(jsonVal);

				} catch (err) {
					skipUpdate=true;
				}

				if(!skipUpdate){
					let jsonString: string = JSON.stringify(jsonVal);

					panel.webview.html = getWebviewContent(
						json5Src,
						jschartingSrc,
						jsonString
					);
				}

			};

			draw();

			panel.onDidDispose(
				() => {
					panel = null;
				},
				null,
				context.subscriptions
			);

			vscode.workspace.onDidChangeTextDocument((ev) => {
				if (ev.document !== document) {
					return;
				}

				draw();
			});
		})
	);

	return {
		extendMarkdownIt(md: any) {
			const highlight = md.options.highlight;

			md.options.highlight = (code: any, lang: any) => {
				if (lang && lang.match(/\bjscharting\b/i)) {
					return `<div class="jscharting">${code.trim()}</div>`;
				}

				return highlight(code, lang);
			};

			return md;
		}
	};
}

function getWebviewContent(
	json5Src: vscode.Uri,
	jschartingSrc: vscode.Uri,
	jschartingConfig: string
) {
	return `<!DOCTYPE html>
		<html lang="en">
		<head>
			<meta charset="UTF-8">
			<meta name="viewport" content="width=device-width, initial-scale=1.0">
			<title>JSCharting</title>
		</head>
		<body>
			<div id="chart"></div>
			<script src="${json5Src}"></script>
			<script src="${jschartingSrc}"></script>
			<script type="text/javascript">
				var chartElement = document.getElementById('chart');
				JSC.Chart(chartElement, ${jschartingConfig});
			</script>
		</body>
		</html>`;
}

export function deactivate() {
}

function fillData(cfg: any) {
	if (!cfg.series || cfg.series.length === 0) {
		cfg.series = [
			{
				name: 'Temporary Data',
				points: [
					{name: 'A', y: 20},
					{name: 'B', y: 50},
					{name: 'C', y: 150},
					{name: 'D', y: 75}
				]
			}
		];
	}
	return cfg;
}
